### PR TITLE
DIRECTOR: Make Wrath of the Gods completable

### PR DESCRIPTION
**Pull request by moralrecordings**

### CODE REVIEW

File: engines/director/lingo/xlibs/wininfo.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include "director/director.h"
+#include "director/lingo/lingo.h"
+#include "director/lingo/lingo-object.h"
+#include "director/lingo/lingo-utils.h"
+#include "director/lingo/xlibs/wininfo.h"
+
+/**************************************************
+ *
+ * USED IN:
+ * Devil's Canyon: A Dinamation Adventure
```
Comment (bluegr) [inline engines/director/lingo/xlibs/wininfo.cpp:33]: And in Wrath of the Gods, as well?

File: engines/director/lingo/xlibs/wininfo.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+static MethodProto xlibMethods[] = {
+	{ "new",				WinInfoXObj::m_new,		 0, 0,	400 },
+	{ "dispose",				WinInfoXObj::m_dispose,		 0, 0,	400 },
+	{ "name",				WinInfoXObj::m_name,		 0, 0,	400 },
+	{ "winInfo",				WinInfoXObj::m_winInfo,		 3, 3,	400 },
+	{ nullptr, nullptr, 0, 0, 0 }
+};
+
+WinInfoXObject::WinInfoXObject(ObjectType ObjectType) :Object<WinInfoXObject>("WinInfoXObj") {
```
Comment (bluegr) [inline engines/director/lingo/xlibs/wininfo.cpp:62]:
```suggestion
WinInfoXObject::WinInfoXObject(ObjectType ObjectType) : Object<WinInfoXObject>("WinInfoXObj") {
```

https://wiki.scummvm.org/index.php?title=Code_Formatting_Conventions#Whitespaces

File: engines/director/lingo/xlibs/wininfo.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		// TODO - Implement Xtra
+	}
+}
+
+void WinInfoXObj::close(int type) {
+	if (type == kXObj) {
+		WinInfoXObject::cleanupMethods();
+		g_lingo->_globalvars[xlibName] = Datum();
+	} else if (type == kXtraObj) {
+		// TODO - Implement Xtra
```
Comment (bluegr) [inline engines/director/lingo/xlibs/wininfo.cpp:81]: Should we throw a warning here?

File: engines/director/lingo/lingo-the.cpp [hunk 4]
```diff
@@ -890,6 +891,10 @@ Datum Lingo::getTheEntity(int entity, Datum &id, int field) {
 	case kTheUpdateMovieEnabled:
 		d = g_lingo->_updateMovieEnabled;
 		break;
+	case kTheVideoForWindowsPresent:
+		// Video For Windows is always present for scummvm
```
Comment (bluegr) [inline engines/director/lingo/lingo-the.cpp:895]:
```suggestion
		// Video For Windows is always present for ScummVM
```

File: engines/director/debugger.cpp [hunk 5]
```diff
@@ -558,8 +605,19 @@ bool Debugger::cmdDisasm(int argc, const char **argv) {
 				debugPrintf("%s\n", lingo->formatFunctionBody(ctx->_functionHandlers[funcName]).c_str());
 				return true;
 			}
+			if (sharedCast->_lingoArchive->factoryContexts.contains(scriptId)) {
```
Comment (bluegr) [inline engines/director/debugger.cpp:608]: This block looks to be a duplicate of the one at line 588. Could both blocks be moved in a reusable function?

File: engines/director/debugger.cpp [hunk 6]
```diff
@@ -521,6 +538,24 @@ bool Debugger::cmdDisasm(int argc, const char **argv) {
 						}
 					}
 				}
+				debugPrintf("  Factories:\n");
```
Comment (bluegr) [inline engines/director/debugger.cpp:541]: This block looks to be a duplicate of the one at line 505. Could both blocks be moved into a reusable function?

File: engines/director/lingo/xlibs/wininfo.cpp (hunk 3 above)
Comment (moralrecordings) [reply engines/director/lingo/xlibs/wininfo.cpp:81]: Xtras are D5 and up and not supported yet, there needs to be additional infra for that to work.

File: engines/director/lingo/xlibs/wininfo.cpp (hunk 1 above)
Comment (moralrecordings) [reply engines/director/lingo/xlibs/wininfo.cpp:33]: This commit is unrelated to the Wrath changes, it's an XOBJ stub from one of the games I was fixing on-stream.

File: engines/director/debugger.cpp (hunk 5 above)
Comment (moralrecordings) [reply engines/director/debugger.cpp]: I've removed the main source of duplication (having to run everything for both the cast and the sharedCast). Most of what's left is looping through the different places script archives are stored and calling formatFunctionBody on them; my preference is to keep the logic for doing so as flat as possible so it's easier to read.

File: engines/director/debugger.cpp (hunk 6 above)
Comment (moralrecordings) [reply engines/director/debugger.cpp]: See above

File: engines/director/lingo/xlibs/wininfo.cpp (hunk 3 above)
Comment (moralrecordings) [reply engines/director/lingo/xlibs/wininfo.cpp:81]: To clarify, we have that stub code in all of the xlib implementations, but it's in anticipation of a rather large architecture feature that hasn't been added yet. D5 apps which rely on Xtras will fail and throw warnings much earlier in the execution.

**Comment by sev-** [comment]
Thanks!

**Pull request merged**